csharp/src/Ice/INetworkProxy.cs

```diff
@@ -192,42 +183,50 @@ public int EndWrite(IList<ArraySegment<byte>> buffer, int bytesTransferred)
             return bytesTransferred < buffer.GetByteCount() ? SocketOperation.Write : SocketOperation.Read;
         }
 
-        public void BeginRead(Buffer buf)
+        public void BeginRead(ref ArraySegment<byte> buffer)
         {
-            //
-            // Read the HTTP response
-            //
-            buf.Resize(7, true); // Enough space for reading at least HTTP1.1
-            buf.B.Position(0);
+            // Read the HTTP response, reserve enough space for reading at
+            // least HTTP1.1
+            buffer = new ArraySegment<byte>(new byte[256], 0, 7);
         }
 
-        public int EndRead(Buffer buf)
+        public int EndRead(ref ArraySegment<byte> buffer, int offset)
         {
             //
             // Check if we received the full HTTP response, if not, continue
             // reading otherwise we're done.
             //
-            int end = new HttpParser().IsCompleteMessage(buf.B, 0, buf.B.Position());
-            if (end < 0 && !buf.B.HasRemaining())
+            int end = HttpParser.IsCompleteMessage(buffer.AsSpan(0, offset));
+            if (end < 0 && offset == buffer.Count)
             {
                 //
                 // Read one more byte, we can't easily read bytes in advance
                 // since the transport implenentation might be be able to read
                 // the data from the memory instead of the socket.
                 //
-                buf.Resize(buf.Size() + 1, true);
+                if (offset == buffer.Array.Length)
```

Comment:
All this code assumes that `buffer.Offset == 0`. You should add an Assert to make this clear.

Reply:
Added an assert